### PR TITLE
Added IPAddress.h SPI.h , WCharacter.h, sqrt and delayMicroseconds

### PR DESCRIPTION
**Pull request by ramboerik**
Hi,

I'm using your linux port as a drop in replacement of the official Arduino core when compiling unit tests. I ran across some missing functions and added those that gave me compiler errors.  Most is just copy and paste from the official core!

### CODE REVIEW

File: Arduino.cpp [hunk 1]
```diff
@@ -21,6 +21,10 @@
 // Arduino methods emulated in Unix
 // -----------------------------------------------------------------------
 
+void delayMicroseconds(unsigned long us) {
```
Comment (bxparks) [inline Arduino.cpp:24]: This should be
```C++
void delayMicroSeconds(unsigned int us)
```
to be compatible with other cores.

File: Arduino.h [hunk 2]
```diff
@@ -81,6 +82,7 @@
 
 extern "C" {
 
+void delayMicroseconds(unsigned long us);
```
Comment (bxparks) [inline Arduino.h:85]: ditto, change `unsigned long` to `unsigned  int`

File: IPAddress.cpp [hunk 3]
```diff
@@ -0,0 +1,113 @@
+/*
+  IPAddress.cpp - Base class that provides IPAddress
+  Copyright (c) 2011 Adrian McEwen.  All right reserved.
+
```
Comment (bxparks) [inline IPAddress.cpp:4]: Add something like:
```
Copied from Arduino AVR core 1.8.3 by Erik Tideman. Removed
unnecessary 'friend' declarations.

File: SPI.cpp [hunk 4]
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (c) 2010 by Cristian Maglie <c.maglie@arduino.cc>
+ * Copyright (c) 2014 by Paul Stoffregen <paul@pjrc.com> (Transaction API)
+ * Copyright (c) 2014 by Matthijs Kooijman <matthijs@stdin.nl> (SPISettings AVR)
+ * Copyright (c) 2014 by Andrew J. Kroll <xxxajk@gmail.com> (atomicity fixes)
+ * SPI Master library for arduino.
```
Comment (bxparks) [inline SPI.cpp:6]: Add something like:
```
Copyright (c) 2021 by Erik Tideman (Replaced actual implementations with stubs.)
```

File: SPI.h [hunk 5]
```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (c) 2010 by Cristian Maglie <c.maglie@arduino.cc>
+ * Copyright (c) 2014 by Paul Stoffregen <paul@pjrc.com> (Transaction API)
+ * Copyright (c) 2014 by Matthijs Kooijman <matthijs@stdin.nl> (SPISettings AVR)
+ * Copyright (c) 2014 by Andrew J. Kroll <xxxajk@gmail.com> (atomicity fixes)
+ * SPI Master library for arduino.
```
Comment (bxparks) [inline SPI.h:6]: Add something like:
```
Copyright (c) 2021 by Erik Tideman (Replaced actual implementations with stubs.)
```

File: WCharacter.h [hunk 6]
```diff
@@ -0,0 +1,168 @@
+/*
+ WCharacter.h - Character utility functions for Wiring & Arduino
+ Copyright (c) 2010 Hernando Barragan.  All right reserved.
+
```
Comment (bxparks) [inline WCharacter.h:4]: Add something like:
```
Copyright (c) 2021 by Erik Tideman (Replaced 'boolean' with 'bool')
```

**Review by bxparks — COMMENTED** [review]
Hi, thanks, this is great. I like how you fixed some style issues for consistency, e.g. trailing white spaces and changing `boolean` to `bool`.

A few requests before I merge this:

1) Can you rebase your PR onto the latest commit on the `develop` branch? There should be no conflicts with anything that I committed since you branched.
2) Can you update the copyright note at the top to add your name, and a few lines to note that you changed a few things?
3) An incorrect function signature on `delayMicroseconds()`, I'll point that out in the review.

**Comment by ramboerik** [comment]
Hi again, 
all comments should be resolved now,

Thanks for great work!

**Comment by bxparks** [comment]
This looks great, thanks for sending the PR.
There is one little compiler warning in SPI.h. You forgot to add a 'return 0' from `uint8_t transfer(uint8_t data)`. I'll fix that after I merge this in.

**Pull request merged**